### PR TITLE
Fix overload resolution regression around params parameters

### LINKED ISSUE #73346: Overload resolution ambiguity introduced by SDK 8.0.300-preview.24203.14
The SDK 8 preview that comes with the latest VS2022 preview introduces an overload resolution ambiguity that is not in the stable SDK.

**Version Used**:

8.0.300-preview.24203.14

**Steps to Reproduce**:

See [this repo](https://github.com/danielcweber/OverloadResolutionRepro) for a reproduction and [this](https://github.com/danielcweber/OverloadResolutionRepro/actions/runs/8719145734) or [this](https://github.com/danielcweber/OverloadResolutionRepro/actions/runs/8720044992) workflow run to watch a build fail.

The issue is also in the latest SDK 9 preview. It's not reproducible on dotnetfiddle/sharplab, the latest SDK bits there are seemingly not recent enough.

Everything is fine on 8.0.204.

This might be related to collection expression conversions since the method overload in question has a `params` [] parameter.

Note: I also opened an issue in [the sdk repo](https://github.com/dotnet/sdk/issues/40235) but this might be the more appropriate place.

### PR DESCRIPTION
**Pull request by AlekseyTs**
Fixes #73346.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs [hunk 1]
```diff
@@ -2261,11 +2245,9 @@ private static ParameterSymbol GetParameter(int argIndex, MemberAnalysisResult r
                         continue;
                     }
 
-                    var parameter1 = GetParameter(i, m1.Result, m1Original);
-                    uninst1.Add(GetParameterType(parameter1, m1.Result, m1Original.Length));
+                    uninst1.Add(getParameterTypeAndRefKind(i, m1.Result, m1DefinitionParameters, m1.Result.DefinitionParamsElementTypeOpt, out _));
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs:2248]: >DefinitionParamsElementTypeOpt [](http://example.com/codeflow?start=107&length=30)

This is where the actual fix is

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs [hunk 2]
```diff
@@ -2261,11 +2245,9 @@ private static ParameterSymbol GetParameter(int argIndex, MemberAnalysisResult r
...
 
-                    var parameter2 = GetParameter(i, m2.Result, m2Original);
-                    uninst2.Add(GetParameterType(parameter2, m2.Result, m2Original.Length));
+                    uninst2.Add(getParameterTypeAndRefKind(i, m2.Result, m2DefinitionParameters, m2.Result.DefinitionParamsElementTypeOpt, out _));
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs:2250]: >DefinitionParamsElementTypeOpt [](http://example.com/codeflow?start=107&length=30)

This is where the actual fix is

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs [hunk 3]
```diff
@@ -2353,6 +2335,26 @@ private static ParameterSymbol GetParameter(int argIndex, MemberAnalysisResult r
             }
 
             return BetterResult.Neither;
+
+            // Returns the parameter type (considering params).
+            static TypeSymbol getParameterTypeAndRefKind(int i, MemberAnalysisResult result, ImmutableArray<ParameterSymbol> parameters, TypeWithAnnotations paramsElementTypeOpt, out RefKind parameter1RefKind)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs:2340]: >paramsElementTypeOpt [](http://example.com/codeflow?start=157&length=20)

Is this always `result.DefinitionParamsElementTypeOpt`? If so, consider removing this parameter.

File: src/Compilers/CSharp/Test/Emit2/Semantics/ParamsCollectionTests.cs [hunk 4]
```diff
@@ -15575,5 +15575,42 @@ static void Main()
...
+        {
+            string source = """
+using System;
+
+namespace OverloadResolutionRepro
+{
+    public class Foo
+    {
+        public void Method<S>(params Func<Bar, S>[] projections) => Console.Write(1);
+        public void Method<S>(params Func<Bar, Wrapper<S>>[] projections) => Console.Write(2);
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit2/Semantics/ParamsCollectionTests.cs:15591]: >Method [](http://example.com/codeflow?start=20&length=6)

Please test a constructor as well.

**Review by cston — APPROVED** [review]

**Comment by AlekseyTs** [comment]
@RikkiGibson, @333fred, @dotnet/roslyn-compiler For the second review

**Comment by AlekseyTs** [comment]
@RikkiGibson, @333fred, @dotnet/roslyn-compiler For the second review

**Review by RikkiGibson — APPROVED** [review]

**Comment by danielcweber** [comment]
We'd please need some kind of workaround because the bug has found its way into 8.0.300, the stable Visual studio, the preview of Visual Studio anyway, uninstalling 8.0.300 won't do, uninstalling 9.x-preview won't do, fixating 8.0.204 in global.json and disallowing rollforward won't do.

**Comment by jjonescz** [comment]
As a temporary workaround you could try installing `Microsoft.Net.Compilers.Toolset` which will force use of a specific compiler version. If you are not relying on new features, you could try using some older version from [NuGet](https://www.nuget.org/packages/Microsoft.Net.Compilers.Toolset); or you can use the latest version from [AzDo feed](https://dev.azure.com/dnceng/public/_artifacts/feed/dotnet-tools/NuGet/Microsoft.Net.Compilers.Toolset/) (but you need to add that feed to a NuGet.config). Don't forget to uninstall this package once the fix is released in the SDK, because it's *not* supported.

**Pull request merged**